### PR TITLE
[5.0] Adapt to changes in bls12-381 lib and add tests

### LINKED ISSUE #1701: Modify BLS host functions to fix failing tests on ARM builds
Depends on https://github.com/AntelopeIO/bls12-381/issues/17.

Some BLS tests, like the garbage tests in `test_fc`, seem to be failing on ARM builds. It's a little surprising because it appears the main bls test is passing. Could be something specific about the garbage tests + C impl. Needs more investigation.

ARM is not supported, but failures there on newly added consensus code are probably worth understanding.

### PR DESCRIPTION
**Pull request by yarkinwho**
Removes host functions:
- `bls_g1_mul`
- `bls_g2_mul`
- `bls_g1_exp`
- `bls_g2_exp`

Adds host functions:
- `bls_g1_weighted_sum`
- `bls_g2_weighted_sum`
- `bls_fp_mul`
- `bls_fp_exp`

All bls host functions modified to accept and return affine little endian form.

Removes protocol feature: `BLS_PRIMITIVES` and adds protocol feature `BLS_PRIMITIVES2`.

Resolves #1701

### CODE REVIEW

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp [hunk 1]
```diff
@@ -1808,41 +1819,42 @@ namespace webassembly {
...
-          * Host function for G1 scalar multiplication on the elliptic curve bls12-381
+          * Host function for G2 addition on the elliptic curve bls12-381
           *
           * @ingroup crypto
-          * @param point - a span containing the G1 point operand.
-          * @param scalar - a span containing the scalar operand.
-          * @param[out] result - the result: scalar * point.
+          * @param op1 - a span containing the first operand G2 point in Montgemory form.
+          * @param op2 - a span containing the second operand G2 point in Montgemory form.
+          * @param[out] result - the result op1 + op2 in Montgemory form.
```
Comment (greg7mdp) [inline libraries/chain/include/eosio/chain/webassembly/interface.hpp:1827]:
```suggestion
          * @param op1 - a span containing the first operand G2 point in Montgomery form.
          * @param op2 - a span containing the second operand G2 point in Montgomery form.
          * @param[out] result - the result op1 + op2 in Montgomery form.
```

This typo occurs in many other places as well, please fix all.

Comment (yarkinwho) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp]: fixed

File: libraries/chain/webassembly/crypto.cpp [hunk 2]
```diff
@@ -20,6 +21,193 @@ namespace {
...
+      std::vector<std::tuple<bls12_381::g1, bls12_381::g2>> v;
+      v.reserve(n);
+      for(uint32_t i = 0; i < n; i++)
+      {
+         std::optional<bls12_381::g1> p_g1 = bls12_381::g1::fromAffineBytesLE(std::span<const uint8_t, 96>((const uint8_t*)g1_points.data() + i*96, 96), true, mont);
+         std::optional<bls12_381::g2> p_g2 = bls12_381::g2::fromAffineBytesLE(std::span<const uint8_t, 192>((const uint8_t*)g2_points.data() + i*192, 192), true, mont);
+         if(!p_g1 || !p_g2)
+            return return_code::failure;
+         bls12_381::pairing::add_pair(v, *p_g1, *p_g2);
+         if(i%4 == 0)
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply libraries/chain/webassembly/crypto.cpp]: This code has been refactored when removing _mont functions. This should no longer be an issue now.

File: libraries/chain/webassembly/crypto.cpp [hunk 3]
```diff
@@ -20,6 +21,193 @@ namespace {
...
+      sv.reserve(n);
+      for(uint32_t i = 0; i < n; i++)
+      {
+         std::optional<bls12_381::g2> p = bls12_381::g2::fromAffineBytesLE(std::span<const uint8_t, 192>((const uint8_t*)points.data() + i*192, 192), true, mont);
+         if(!p)
+            return return_code::failure;
+         std::array<uint64_t, 4> s = bls12_381::scalar::fromBytesLE<4>(std::span<const uint8_t, 32>((const uint8_t*)scalars.data() + i*32, 32));
+         pv.push_back(*p);
+         sv.push_back(s);
+         if(i%6 == 0)
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply libraries/chain/webassembly/crypto.cpp]: This code has been refactored when removing _mont functions. This should no longer be an issue now.

File: libraries/chain/webassembly/crypto.cpp [hunk 4]
```diff
@@ -20,6 +21,193 @@ namespace {
...
+      sv.reserve(n);
+      for(uint32_t i = 0; i < n; i++)
+      {
+         std::optional<bls12_381::g1> p = bls12_381::g1::fromAffineBytesLE(std::span<const uint8_t, 96>((const uint8_t*)points.data() + i*96, 96), true, mont);
+         if(!p.has_value())
+            return return_code::failure;
+         std::array<uint64_t, 4> s = bls12_381::scalar::fromBytesLE<4>(std::span<const uint8_t, 32>((const uint8_t*)scalars.data() + i*32, 32));
+         pv.push_back(p.value());
+         sv.push_back(s);
+         if(i%10 == 0)
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply libraries/chain/webassembly/crypto.cpp]: This code has been refactored when removing _mont functions. This should no longer be an issue now.

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp [hunk 5]
```diff
@@ -1865,6 +1889,18 @@ namespace webassembly {
...
          int32_t bls_pairing(span<const char> g1_points, span<const char> g2_points, const uint32_t n, span<char> result) const;
 
+         /**
+          * Host function to calculate the pairing of (G1, G2) pairs on the elliptic curve bls12-381
+          *
+          * @ingroup crypto
+          * @param g1_points - a span containing a list of G1 points (P0, P1, P2... Pn) in Montgomery form.
+          * @param g2_points - a span containing a list of G2 points (P0, P1, P2... Pn) in Montgomery form.
+          * @param n - the number of elements in the lists.
+          * @param[out] result - the result e(g1_0, g2_0) * e(g1_1, g2_1) * ... * e(g1_n, g2_n) in Montgomery form.
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp]: Updated

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp [hunk 6]
```diff
@@ -1851,7 +1863,19 @@ namespace webassembly {
...
+
+         /**
+          * Host function for G2 weighted sum (multi-exponentiation) on the elliptic curve bls12-381
+          *
+          * @ingroup crypto
+          * @param points - a span containing a list of G2 points (P0, P1, P2... Pn) in Montgomery form.
+          * @param scalars - a span containing a list of scalars (s0, s1, s2... sn).
+          * @param n - the number of elements in the lists.
+          * @param[out] result - the result s0 * P0 + s1 * P1 + ... + sn * Pn in Montgomery form.
+          * @return -1 if there was an error 0 otherwise
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp]: Updated

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp [hunk 7]
```diff
@@ -1808,41 +1819,42 @@ namespace webassembly {
...
-          * Host function for G1 scalar multiplication on the elliptic curve bls12-381
+          * Host function for G2 addition on the elliptic curve bls12-381
           *
           * @ingroup crypto
-          * @param point - a span containing the G1 point operand.
-          * @param scalar - a span containing the scalar operand.
-          * @param[out] result - the result: scalar * point.
+          * @param op1 - a span containing the first operand G2 point in Montgomery form.
+          * @param op2 - a span containing the second operand G2 point in Montgomery form.
+          * @param[out] result - the result op1 + op2 in Montgomery form.
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp]: Updated

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp [hunk 8]
```diff
@@ -1796,6 +1796,17 @@ namespace webassembly {
...
          int32_t bls_g1_add(span<const char> op1, span<const char> op2, span<char> result) const;
 
+         /**
+          * Host function for G1 addition on the elliptic curve bls12-381
+          *
+          * @ingroup crypto
+          * @param op1 - a span containing the first operand G1 point in Montgomery form.
+          * @param op2 - a span containing the second operand G1 point in Montgomery form.
+          * @param[out] result - the result op1 + op2 in Montgomery form.
+          * @return -1 if there was an error 0 otherwise
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp]: Updated

File: benchmark/bls.cpp [hunk 9]
```diff
@@ -406,14 +406,14 @@ void benchmark_bls_fp_mod() {
 void bls_benchmarking() {
    benchmark_bls_g1_add();
    benchmark_bls_g2_add();
-   benchmark_bls_g1_mul();
-   benchmark_bls_g2_mul();
+   // benchmark_bls_g1_mul();
+   // benchmark_bls_g2_mul();
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply benchmark/bls.cpp]: removed

File: benchmark/bls.cpp [hunk 10]
```diff
@@ -162,13 +162,13 @@ void benchmark_bls_g2_add() {
 
    benchmarking("bls_g2_add", benchmarked_func);
 }
-
+/*
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply benchmark/bls.cpp]: removed

**Comment by greg7mdp** [comment]
I'm not quite understanding the logic of this function:

```
array<uint8_t, 48> fp::toBytesBE(const bool raw) const
{
    if(raw) return scalar::toBytesBE<6>(d);
    else    return scalar::toBytesBE<6>(fromMont().d);
}
```

In particular what does `raw` mean. Does it mean, as I understand, that `*this` is already in Montgomery form?

Is that is the case, when `raw` is false, and `*this` is *not* in Montgomery form, Why do we call `fromMont()`? Shouldn't it be `toMont()`?

**Comment by greg7mdp** [comment]
There are inconsistencies between the types used in the <span> for various APIs. 
* Sometimes we use `uint8_t` (which translates to `unsigned char`, as in `void g1::toAffineBytesBE(const span<uint8_t, 96> out, const bool raw) const`.
* Sometimes we use `char`, as in `interface::bls_g1_add(span<const char> op1, span<const char> op2, span<char> result) const`.

Can we use the same type throughout?

This would help making the code simpler and avoid unnecessary repetition of sizes, which can be error-prone. 
For example, instead of:

```c++
void benchmark_bls_g1_add() {
   // prepare g1 operand in Jacobian LE format
   g1 p = random_g1();
   std::vector<char> buf(96);
   p.toAffineBytesLE(std::span<uint8_t, 96>((uint8_t*)buf.data(), 96), true);
   eosio::chain::span<const char> op1(buf.data(), buf.size());

   // prepare result operand
   std::vector<char> result_buf(96);
   eosio::chain::span<char> result(result_buf.data(), result_buf.size());

   // set up bls_g1_add to be benchmarked
   interface_in_benchmark interface;
   auto g1_add_func = [&]() {
      interface.interface->bls_g1_add(op1, op1, result);
   };

   benchmarking("bls_g1_add", g1_add_func);
}
```

we could have:

```c++
void benchmark_bls_g1_add() {
   // prepare g1 operand in Jacobian LE format
   g1 p = random_g1();
   std::array<unsigned char, 96> op1;
   p.toAffineBytesLE(op1, true);

   // prepare result operand
   std::array<char, 96> result;

   // set up bls_g1_add to be benchmarked
   interface_in_benchmark interface;
   auto g1_add_func = [&]() {
      interface.interface->bls_g1_add(op1, op1, result);
   };

   benchmarking("bls_g1_add", g1_add_func);
}
```

**Comment by yarkinwho** [comment]
> I'm not quite understanding the logic of this function:
> 
> ```
> array<uint8_t, 48> fp::toBytesBE(const bool raw) const
> {
>     if(raw) return scalar::toBytesBE<6>(d);
>     else    return scalar::toBytesBE<6>(fromMont().d);
> }
> ```
> 
> In particular what does `raw` mean. Does it mean, as I understand, that `*this` is already in Montgomery form?
> 
> Is that is the case, when `raw` is false, and `*this` is _not_ in Montgomery form, Why do we call `fromMont()`? Shouldn't it be `toMont()`?

raw here means whether the the output should be the "raw" form used for internal calculation, aka. Montgomery form. 
*this is supposed to always be in Montgomery form. That's why we need extra caution when constructing fp.



.

**Comment by yarkinwho** [comment]
> in fp.cpp, it looks like we check `isValid()` only if the `check` flag is set, even though the comment says otherwise:
> 
> ```
> optional<fp> fp::fromBytesBE(const span<const uint8_t, 48> in, const bool check, const bool raw)
> {
>     // We decided to always validate the input here. But we reserve the flag.
>     fp e = fp(scalar::fromBytesBE<6>(in));
>     if(check && !e.isValid()) return {};
>     if(raw) return e;
>     else    return e.toMont();
> }
> 
> optional<fp> fp::fromBytesLE(const span<const uint8_t, 48> in, const bool check, const bool raw)
> {
>     // We decided to always validate the input here. But we reserve the flag.
>     fp e = fp(scalar::fromBytesLE<6>(in));
>     if(check && !e.isValid()) return {};
>     if(raw) return e;
>     else    return e.toMont();
> }
> ```

Oh, that's the comment I forgot to delete...

**Comment by yarkinwho** [comment]
> There are inconsistencies between the types used in the for various APIs.
> 
> * Sometimes we use `uint8_t` (which translates to `unsigned char`, as in `void g1::toAffineBytesBE(const span<uint8_t, 96> out, const bool raw) const`.
> * Sometimes we use `char`, as in `interface::bls_g1_add(span<const char> op1, span<const char> op2, span<char> result) const`.
> 
> Can we use the same type throughout?
> 
> Also use `std::array` instead of `std::vector` when appropriate, as it converts implicitely to `eosio::chain::span`.
> 
> This would help making the code simpler and avoid unnecessary repetition of sizes, which can be error-prone. For example, instead of:
> 
> ```c++
> void benchmark_bls_g1_add() {
>    // prepare g1 operand in Jacobian LE format
>    g1 p = random_g1();
>    std::vector<char> buf(96);
>    p.toAffineBytesLE(std::span<uint8_t, 96>((uint8_t*)buf.data(), 96), true);
>    eosio::chain::span<const char> op1(buf.data(), buf.size());
> 
>    // prepare result operand
>    std::vector<char> result_buf(96);
>    eosio::chain::span<char> result(result_buf.data(), result_buf.size());
> 
>    // set up bls_g1_add to be benchmarked
>    interface_in_benchmark interface;
>    auto g1_add_func = [&]() {
>       interface.interface->bls_g1_add(op1, op1, result);
>    };
> 
>    benchmarking("bls_g1_add", g1_add_func);
> }
> ```
> 
> we could have:
> 
> ```c++
> void benchmark_bls_g1_add() {
>    // prepare g1 operand in Jacobian LE format
>    g1 p = random_g1();
>    std::array<unsigned char, 96> op1;
>    p.toAffineBytesLE(op1, true);
> 
>    // prepare result operand
>    std::array<char, 96> result;
> 
>    // set up bls_g1_add to be benchmarked
>    interface_in_benchmark interface;
>    auto g1_add_func = [&]() {
>       interface.interface->bls_g1_add(op1, op1, result);
>    };
> 
>    benchmarking("bls_g1_add", g1_add_func);
> }
> ```

good points. guess that's because that's code originally written by different people :) 
Let me try if i can unify them

**Comment by greg7mdp** [comment]
> > I'm not quite understanding the logic of this function:
> > ```
> > array<uint8_t, 48> fp::toBytesBE(const bool raw) const
> > {
> >     if(raw) return scalar::toBytesBE<6>(d);
> >     else    return scalar::toBytesBE<6>(fromMont().d);

> raw here means whether the the output should be the "raw" form used for internal calculation, aka. Montgomery form. *this is supposed to always be in Montgomery form. That's why we need extra caution when constructing fp.

So when we call `scalar::toBytesBE<6>()`, would it not work if we used the Montgomery form as input?

**Comment by yarkinwho** [comment]
> > > I'm not quite understanding the logic of this function:
> > > ```
> > > array<uint8_t, 48> fp::toBytesBE(const bool raw) const
> > > {
> > >     if(raw) return scalar::toBytesBE<6>(d);
> > >     else    return scalar::toBytesBE<6>(fromMont().d);
> > > ```
> 
> > raw here means whether the the output should be the "raw" form used for internal calculation, aka. Montgomery form. *this is supposed to always be in Montgomery form. That's why we need extra caution when constructing fp.
> 
> So when we call `scalar::toBytesBE<6>()`, would it not work if we used the Montgomery form as input?

scalar is defined as numbers that are not in GF(p) so there's no Montgomery form.
So:
 if(raw) return scalar::toBytesBE<6>(d); <-we want to export the "raw" number, so treat it as scalar and export.
else    return scalar::toBytesBE<6>(fromMont().d); <- we want to export the "not raw" number so it will need to be converted back "fromMont"

**Comment by greg7mdp** [comment]
A suggestion, do you think it might be better to define the sizes as members instead of having the hardcoded sizes everywhere, as in:

```c++
class g1
{

public:
    fp x;
    fp y;
    fp z;

    static constexpr size_t jacobian_sz   = 144;
    static constexpr size_t affine_sz     = 96;
    static constexpr size_t compressed_sz = 48;
    
    g1();
    explicit g1(const std::array<fp, 3>& e3);
    g1(const g1& e);
    static std::optional<g1> fromJacobianBytesBE(const std::span<const uint8_t, jacobian_sz> in, const bool check = false, const bool raw = false);
...
```

**Comment by yarkinwho** [comment]
> A suggestion, do you think it might be better to define the sizes as members instead of having the hardcoded sizes everywhere. We could also define buffer types:
> 
> ```c++
> class g1
> {
> 
> public:
>     fp x;
>     fp y;
>     fp z;
> 
>     static constexpr size_t jacobian_sz   = 144;
>     static constexpr size_t affine_sz     = 96;
>     static constexpr size_t compressed_sz = 48;
> 
>     using jacobian_buf   = std::array<uint8_t, jacobian_sz>;
>     using affine_buf     = std::array<uint8_t, affine_sz>;
>     using compressed_buf = std::array<uint8_t, compressed_sz>;
>     
>     g1();
>     explicit g1(const std::array<fp, 3>& e3);
>     g1(const g1& e);
>     static std::optional<g1> fromJacobianBytesBE(const std::span<const uint8_t, jacobian_sz> in, const bool check = false, const bool raw = false);
> ...
> ```

Not sure, usually it's not good to have magic numbers around, but in this case, maybe it's more clear to have those numbers in function declaration so that people don't need to click around to find the correct size.

**Comment by greg7mdp** [comment]
> Not sure, usually it's not good to have magic numbers around, but in this case, maybe it's more clear to have those numbers in function declaration so that people don't need to click around to find the correct size.

Hum, I don't know, maybe @arhag can give his opinion.

I think I'd even define `span` types, so we'd have something like this:

```c++
class g1
{

public:
    fp x;
    fp y;
    fp z;

    static constexpr size_t jacobian_sz   = 144;
    static constexpr size_t affine_sz     = 96;
    static constexpr size_t compressed_sz = 48;

    using jacobian_buf   = std::array<uint8_t, jacobian_sz>;
    using affine_buf     = std::array<uint8_t, affine_sz>;
    using compressed_buf = std::array<uint8_t, compressed_sz>;

    using jacobian_in    = std::span<const uint8_t, jacobian_sz>;
    using jacobian_out   = std::span<uint8_t, jacobian_sz>;
    using affine_in      = std::span<const uint8_t, affine_sz>;
    using affine_out     = std::span<uint8_t, affine_sz>;
    using compressed_in  = std::span<const uint8_t, compressed_sz>;
    using compressed_out = std::span<uint8_t, compressed_sz>;
    
    
    g1();
    explicit g1(const std::array<fp, 3>& e3);
    g1(const g1& e);
    static std::optional<g1> fromJacobianBytesBE(const jacobian_in in, const bool check = false, const bool raw = false);
    static std::optional<g1> fromJacobianBytesLE(const jacobian_in in, const bool check = false, const bool raw = false);
    static std::optional<g1> fromAffineBytesBE(const affine_in in, const bool check = false, const bool raw = false);
    static std::optional<g1> fromAffineBytesLE(const affine_in in, const bool check = false, const bool raw = false);
    static std::optional<g1> fromCompressedBytesBE(const compressed_in in);
    void toJacobianBytesBE(const jacobian_out out, const bool raw = false) const;
    void toJacobianBytesLE(const jacobian_out out, const bool raw = false) const;
    void toAffineBytesBE(const affine_out out, const bool raw = false) const;
    void toAffineBytesLE(const affine_out out, const bool raw = false) const;
    void toCompressedBytesBE(const compressed_out out) const;
    jacobian_buf toJacobianBytesBE(const bool raw = false) const;
    jacobian_buf toJacobianBytesLE(const bool raw = false) const;
    affine_buf toAffineBytesBE(const bool raw = false) const;
    affine_buf toAffineBytesLE(const bool raw = false) const;
    compressed_buf toCompressedBytesBE() const;
```

**Comment by yarkinwho** [comment]
> > Not sure, usually it's not good to have magic numbers around, but in this case, maybe it's more clear to have those numbers in function declaration so that people don't need to click around to find the correct size.
> 
> Hum, I don't know, maybe @arhag can give his opinion.
> 
> I think I'd even define `span` types, so we'd have something like this:
> 
> ```c++
> class g1
> {
> 
> public:
>     fp x;
>     fp y;
>     fp z;
> 
>     static constexpr size_t jacobian_sz   = 144;
>     static constexpr size_t affine_sz     = 96;
>     static constexpr size_t compressed_sz = 48;
> 
>     using jacobian_buf   = std::array<uint8_t, jacobian_sz>;
>     using affine_buf     = std::array<uint8_t, affine_sz>;
>     using compressed_buf = std::array<uint8_t, compressed_sz>;
> 
>     using jacobian_in    = std::span<const uint8_t, jacobian_sz>;
>     using jacobian_out   = std::span<uint8_t, jacobian_sz>;
>     using affine_in      = std::span<const uint8_t, affine_sz>;
>     using affine_out     = std::span<uint8_t, affine_sz>;
>     using compressed_in  = std::span<const uint8_t, compressed_sz>;
>     using compressed_out = std::span<uint8_t, compressed_sz>;
>     
>     
>     g1();
>     explicit g1(const std::array<fp, 3>& e3);
>     g1(const g1& e);
>     static std::optional<g1> fromJacobianBytesBE(const jacobian_in in, const bool check = false, const bool raw = false);
>     static std::optional<g1> fromJacobianBytesLE(const jacobian_in in, const bool check = false, const bool raw = false);
>     static std::optional<g1> fromAffineBytesBE(const affine_in in, const bool check = false, const bool raw = false);
>     static std::optional<g1> fromAffineBytesLE(const affine_in in, const bool check = false, const bool raw = false);
>     static std::optional<g1> fromCompressedBytesBE(const compressed_in in);
>     void toJacobianBytesBE(const jacobian_out out, const bool raw = false) const;
>     void toJacobianBytesLE(const jacobian_out out, const bool raw = false) const;
>     void toAffineBytesBE(const affine_out out, const bool raw = false) const;
>     void toAffineBytesLE(const affine_out out, const bool raw = false) const;
>     void toCompressedBytesBE(const compressed_out out) const;
>     jacobian_buf toJacobianBytesBE(const bool raw = false) const;
>     jacobian_buf toJacobianBytesLE(const bool raw = false) const;
>     affine_buf toAffineBytesBE(const bool raw = false) const;
>     affine_buf toAffineBytesLE(const bool raw = false) const;
>     compressed_buf toCompressedBytesBE() const;
> ```
> 
> In user code, I think this:
> 
> ```c++
>    g2::affine_buf buf;
>    random_g2().toAffineBytesLE(buf, true);
> ```
> 
> is better than this:
> 
> ```c++
>    std::vector<char> buf(192);
>    random_g2().toAffineBytesLE(std::span<uint8_t, 192>((uint8_t*)buf.data(), 192), true);
> ```

I guess if we are going to pick similar approaches, we will need some better design for hexToBytes and bytesToHex methods. With the current design, I think we will still need to put some hardcoded number to copy data to buf in some steps.

File: libraries/chain/protocol_feature_manager.cpp [hunk 11]
```diff
@@ -263,12 +263,13 @@ Enables new `get_block_num` intrinsic which returns the current block number.
...
             "BLS_PRIMITIVES",
-            fc::variant("01969c44de35999b924095ae7f50081a7f274409fdbccb9fc54fa7836c76089c").as<digest_type>(),
+            fc::variant("dbb957731e310f86462d1643fd98f69eedb193892e3c1d38593410e34ad5d8ef").as<digest_type>(),
             // SHA256 hash of the raw message below within the comment delimiters (do not modify message below).
 /*
 Builtin protocol feature: BLS_PRIMITIVES
+
 Adds new cryptographic host functions
-- Add, multiply, multi-exponentiation and pairing functions for the bls12-381 elliptic curve.
+- Add, weighted-sum, map, and pairing functions for the bls12-381 elliptic curve.
```
Comment (heifner) [inline libraries/chain/protocol_feature_manager.cpp:272]: https://t.me/jungletestnet/87773

Test restarting a node that had the previous digest for BLS_PRIMITIVES. Consider renaming if it avoid the restart issue.

Comment (yarkinwho) [reply libraries/chain/protocol_feature_manager.cpp]: updated

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp [hunk 12]
```diff
@@ -1781,121 +1781,120 @@ namespace webassembly {
...
+          * @param op1 - a span containing the affine coordingates of the first operand G2 point - 192 bytes little-endian.
+          * @param op2 - a span containing the affine coordingates of the second operand G2 point - 192 bytes little-endian.
+          * @param[out] result - the result op1 + op2 - 192 bytes little-endian.
+          * @return -1 if there was an error, 0 otherwise
          */
          int32_t bls_g2_add(span<const char> op1, span<const char> op2, span<char> result) const;
 
          /**
-          * Host function for G1 scalar multiplication on the elliptic curve bls12-381
+          * Host function for G1 weighted sum (multi-exponentiation) on the elliptic curve bls12-381
```
Comment (arhag) [inline libraries/chain/include/eosio/chain/webassembly/interface.hpp:1811]: Remove `(multi-exponentiation)` from comment.

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp [hunk 13]
```diff
@@ -1781,121 +1781,120 @@ namespace webassembly {
...
           * @param[out] result - the result s0 * P0 + s1 * P1 + ... + sn * Pn.
-          * @return -1 if there was an error 0 otherwise
+          * @return -1 if there was an error, 0 otherwise
          */
-         int32_t bls_g1_exp(span<const char> points, span<const char> scalars, const uint32_t n, span<char> result) const;
+         int32_t bls_g1_weighted_sum(span<const char> points, span<const char> scalars, const uint32_t n, span<char> result) const;
 
          /**
-          * Host function for G2 multi-exponentiation on the elliptic curve bls12-381
+          * Host function for G2 weighted sum (multi-exponentiation) on the elliptic curve bls12-381
```
Comment (arhag) [inline libraries/chain/include/eosio/chain/webassembly/interface.hpp:1823]: Remove `(multi-exponentiation)` from comment.

**Review by arhag — COMMENTED** [review]
Also update bls12-381 library submodule to the commit https://github.com/AntelopeIO/bls12-381/commits/13888087ac9225b1f266fee8b149c4effe10fabb.

Otherwise, looks good to me.

**Review by arhag — APPROVED** [review]

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp [hunk 14]
```diff
@@ -1781,121 +1781,120 @@ namespace webassembly {
...
          /**
           * Host function for G1 addition on the elliptic curve bls12-381
           *
           * @ingroup crypto
-          * @param op1 - a span containing the first operand G1 point.
-          * @param op2 - a span containing the second operand G1 point.
-          * @param[out] result - the result op1 + op2.
-          * @return -1 if there was an error 0 otherwise
+          * @param op1 - a span containing the affine coordingates of the first operand G1 point - 96 bytes little-endian.
+          * @param op2 - a span containing the affine coordingates of the second operand G1 point - 96 bytes little-endian.
```
[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp:1793]: Fixed in #1938

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp [hunk 15]
```diff
@@ -1781,121 +1781,120 @@ namespace webassembly {
...
          /**
           * Host function for G2 addition on the elliptic curve bls12-381
           *
           * @ingroup crypto
-          * @param op1 - a span containing the first operand G2 point.
-          * @param op2 - a span containing the second operand G2 point.
-          * @param[out] result - the result op1 + op2.
-          * @return -1 if there was an error 0 otherwise
+          * @param op1 - a span containing the affine coordingates of the first operand G2 point - 192 bytes little-endian.
+          * @param op2 - a span containing the affine coordingates of the second operand G2 point - 192 bytes little-endian.
```
[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp:1804]: Fixed in #1938

**Pull request merged**